src/python/pants/backend/typescript/tsconfig.py

```diff
@@ -0,0 +1,141 @@
+# Copyright 2024 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import json
+import os
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import Iterable, Literal
+
+from pants.engine.collection import Collection
+from pants.engine.fs import DigestContents, FileContent, PathGlobs
+from pants.engine.internals.selectors import Get, concurrently
+from pants.engine.intrinsics import directory_digest_to_digest_contents, path_globs_to_digest
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.util.frozendict import FrozenDict
+
+
+@dataclass(frozen=True)
+class TSConfig:
+    """Parsed tsconfig.json fields with `extends` substitution applied."""
+
+    path: str
+    module_resolution: str | None = None
+    paths: FrozenDict[str, tuple[str, ...]] | None = None
+    base_url: str | None = None
+    extends: str | None = None
+
+    @classmethod
+    def parse_from_content(cls, content: FileContent) -> TSConfig:
+        parsed_ts_config_json = FrozenDict.deep_freeze(json.loads(content.content))
+        compiler_options = parsed_ts_config_json.get("compilerOptions", FrozenDict())
+        return TSConfig(
+            content.path,
+            module_resolution=compiler_options.get("moduleResolution"),
+            paths=compiler_options.get("paths"),
+            base_url=compiler_options.get("baseUrl"),
+            extends=compiler_options.get("extends"),
+        )
+
+    @property
+    def resolution_root_dir(self) -> str:
+        directory = os.path.dirname(self.path)
+        return os.path.join(directory, self.base_url) if self.base_url else directory
+
+
+class AllTSConfigs(Collection[TSConfig]):
+    pass
+
+
+@dataclass(frozen=True)
+class ParseTSConfigRequest:
+    content: FileContent
+    others: DigestContents
+    target_file: Literal["tsconfig.json", "jsconfig.json"]
```

Comment:
This file is all phrased in terms of typescript and "tsconfig"... but jsconfig is also supported. Do you have some background you can explain about why we're framing it as a TS feature despite seeming to also be applied to 'base' JS?

Reply:
microsoft made typescript, then a tsconfig.json.

They added bundler support to tsconfig.json.

They made vscode, wanted intellisense support for bundled javascript just like typescript.

So they introduced jsconfig.json, (which is a tsconfig.json but named differently) to let vscode intellisense.

Quote from https://code.visualstudio.com/docs/languages/jsconfig:
>jsconfig.json is a descendant of [tsconfig.json](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html), which is a configuration file for TypeScript. jsconfig.json is tsconfig.json with "allowJs" attribute set to true.

I suspect most javascript devs use vscode, so most projects have a jsconfig.json (or it isn't that big of an ask to add one). We can do the same thing microsoft did, which is re-use logic, parsing and stuff between javascript and typescript :nerd_face:

Reply:
Bundlers in this context are tools like webpack, parcel, esbuild, rollup.

Reply:
Can also add that my pycharm pro (so I guess webstorm also included) also support jsconfig.json.

Reply:
Ah, cool, thanks for the background. Maybe we could add a brief comment somewhere here with a summary of that? Maybe as a module level doc string or on one of these literals?